File: 2500/2512.py
n=int(input())
D=[*map(int,input().split())]
m=int(input())
Min,Max=0,max(D)
while Min<=Max:
    mid=(Min+Max)//2
    ans=0
    for i in range(n):
        ans+=min(mid,D[i])
    if ans<=m:Min=mid+1
    else:Max=mid-1
print((Min+Max)//2)

File: 2500/2512.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
int D[10000]={0,};

int main(){
    cin >> n;
    int Min = 0, Max = 0;
    for(int i = 0; i < n; i++){
        cin >> D[i];
        if (D[i]>Max) Max=D[i];
    }
    cin >> m;
    while(Min<=Max){
        int mid = (Min+Max)/2;
        long long ans = 0;
        for(int i = 0; i<n; i++){
            ans+=min(mid,D[i]);
        }
        if(ans<=m) Min=mid+1;
        else Max=mid-1;
    }
    cout << (Min+Max)/2 << endl;
}